src/ast/enrichment/annotate_params.c

```diff
@@ -0,0 +1,68 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "../ast.h"
+#include "../../util/arr.h"
+#include "../../query_ctx.h"
+//------------------------------------------------------------------------------
+//  Annotation context - query parameters
+//------------------------------------------------------------------------------
+
+static AnnotationCtx *_AST_NewParamsContext(void) {
+	AnnotationCtx *param_crx = cypher_ast_annotation_context();
+	return param_crx;
+}
+
+static void _collect_query_params_map(const cypher_astnode_t *ast_exp, rax *parameter_map) {
+	cypher_astnode_type_t type = cypher_astnode_type(ast_exp);
+	// In case of identifier.
+	if(type == CYPHER_AST_PARAMETER) {
+		const char *identifier = cypher_ast_parameter_get_name(ast_exp);
+		const cypher_astnode_t **exp_arr = raxFind(parameter_map, (unsigned char *)identifier,
+												   strlen(identifier));
+		// Use array in case of multiple projections for the same named path.
+		if(exp_arr == raxNotFound) exp_arr = array_new(const cypher_astnode_t *, 1);
+		exp_arr = array_append(exp_arr, ast_exp);
+		raxInsert(parameter_map, (unsigned char *)identifier, strlen(identifier), (void *)exp_arr, NULL);
+	} else {
+		// Recurse over children.
+		uint child_count = cypher_astnode_nchildren(ast_exp);
+		for(uint i = 0; i < child_count; i++) {
+			const cypher_astnode_t *child = cypher_astnode_get_child(ast_exp, i);
+			// Recursively continue mapping.
+			_collect_query_params_map(child, parameter_map);
+		}
+	}
+}
+
+static void _annotate_params(AST *ast) {
+	// Check for number of parameters, if there aren't any, return;
+	rax *params_values = QueryCtx_GetParams();
+	if(raxSize(params_values) == 0) return;
+	rax *query_params_map = raxNew();
+	_collect_query_params_map(ast->root, query_params_map);
+	raxIterator iter;
+	raxStart(&iter, query_params_map);
+	raxSeek(&iter, "^", NULL, 0);
+	while(raxNext(&iter)) {
+		const char *key = (const char *)iter.key;
+		const cypher_astnode_t **exp_arr = iter.data;
+		const cypher_astnode_t *param_value = raxFind(params_values, (unsigned char *) key, iter.key_len);
+		if(param_value != raxNotFound) {
```

Comment:
This should always be captured in validations, right? I would replace the if condition with an assertion.

Reply:
fixe